modules/index/src/test/java/org/apache/ignite/internal/index/IndexAvailabilityControllerTest.java

```diff
@@ -0,0 +1,405 @@
+                    metaStorageManager.remove(ByteArray.fromString(partitionBuildIndexKey(indexId, partitionId))),
+                    willCompleteSuccessfully()
+            );
+        }
+
+        awaitActualMetastoreRevision();
+
+        assertStartBuildIndexKeyAbsent(indexId);
+
+        for (int partitionId = 0; partitionId < partitions; partitionId++) {
+            assertPartitionBuildIndexKeyAbsent(indexId, partitionId);
+        }
+
+        assertFalse(indexDescriptor(INDEX_NAME).writeOnly());
+    }
+
+    @Test
+    void testMetastoreKeysAfterDropIndex() throws Exception {
+        createIndex(INDEX_NAME);
+
+        int indexId = indexId(INDEX_NAME);
+
+        dropIndex(INDEX_NAME);
+
+        awaitActualMetastoreRevision();
+
+        assertStartBuildIndexKeyAbsent(indexId);
+
+        for (int partitionId = 0; partitionId < partitions; partitionId++) {
+            assertPartitionBuildIndexKeyAbsent(indexId, partitionId);
+        }
+    }
+
+    @Test
+    void testMetastoreKeysAfterDropIndexForOnlyOnePartition() throws Exception {
+        changePartitionCountInCatalog(1);
+
+        testMetastoreKeysAfterDropIndex();
+    }
+
+    @Test
+    void testMetastoreKeysAfterFinishBuildingOnePartitionAndDropIndex() throws Exception {
+        createIndex(INDEX_NAME);
+
+        int indexId = indexId(INDEX_NAME);
+
+        finishBuildingIndexForPartition(indexId, 0);
+
+        dropIndex(INDEX_NAME);
+
+        awaitActualMetastoreRevision();
+
+        assertStartBuildIndexKeyAbsent(indexId);
+
+        for (int partitionId = 0; partitionId < partitions; partitionId++) {
+            assertPartitionBuildIndexKeyAbsent(indexId, partitionId);
+        }
+    }
+
+    @Test
+    void testMetastoreKeysAfterDropIndexWithLastRemainingPartition() throws Exception {
+        createIndex(INDEX_NAME);
+
+        int indexId = indexId(INDEX_NAME);
+
+        for (int partitionId = 1; partitionId < partitions; partitionId++) {
+            finishBuildingIndexForPartition(indexId, partitionId);
+        }
+
+        dropIndex(INDEX_NAME);
+
+        awaitActualMetastoreRevision();
+
+        assertStartBuildIndexKeyAbsent(indexId);
+
+        for (int partitionId = 0; partitionId < partitions; partitionId++) {
+            assertPartitionBuildIndexKeyAbsent(indexId, partitionId);
+        }
+    }
+
+    @Test
+    void testMetastoreKeysAfterDropIndexWithTwoRemainingPartition() throws Exception {
+        createIndex(INDEX_NAME);
+
+        int indexId = indexId(INDEX_NAME);
+
+        for (int partitionId = 2; partitionId < partitions; partitionId++) {
+            finishBuildingIndexForPartition(indexId, partitionId);
+        }
+
+        dropIndex(INDEX_NAME);
+
+        awaitActualMetastoreRevision();
+
+        assertStartBuildIndexKeyAbsent(indexId);
+
+        for (int partitionId = 0; partitionId < partitions; partitionId++) {
+            assertPartitionBuildIndexKeyAbsent(indexId, partitionId);
+        }
+    }
+
```

Comment:
How about a test for dropping an index after it has been fully built, but has more than one partition?

Reply:
I didn't understand the test scenario a bit.
If the index is built, then all partition keys must be deleted.